File: src/main/java/com/paragon/client/systems/module/Module.java
package com.paragon.client.systems.module;

import com.paragon.Paragon;
import com.paragon.api.event.client.ModuleToggleEvent;
import com.paragon.api.util.Wrapper;
import com.paragon.client.systems.feature.Feature;
import com.paragon.client.systems.module.hud.impl.ArrayListHUD;
import com.paragon.client.systems.module.setting.Bind;
import com.paragon.client.systems.ui.animation.Animation;
import com.paragon.client.systems.module.setting.Setting;
import net.minecraftforge.common.MinecraftForge;
import org.lwjgl.input.Keyboard;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

public class Module extends Feature implements Wrapper {

    // The category the module is in
    private final Category category;

    // Whether the module is visible in the Array List or not
    private final Setting<Boolean> visible = new Setting<>("Visible", true)
            .setDescription("Whether the module is visible in the array list or not");

    // Whether the module is constantly enabled or not
    private final boolean constant = getClass().isAnnotationPresent(Constant.class);

    // Whether the module is ignored by notifications
    private final boolean ignoredByNotifications = getClass().isAnnotationPresent(IgnoredByNotifications.class);

    // Module Settings
    private final List<Setting<?>> settings = new ArrayList<>();
    private final Setting<Bind> bind = new Setting<>("Bind", new Bind(Keyboard.KEY_NONE, Bind.Device.KEYBOARD))
            .setDescription("The keybind of the module");

    // Arraylist animation
    public Animation animation = new Animation(() -> ArrayListHUD.animationSpeed.getValue(), false, ArrayListHUD.easing::getValue);

    // Whether the module is enabled
    private boolean enabled;

    public Module(String name, Category category, String description) {
        super(name, description);
        this.category = category;

        if (constant) {
            this.enabled = true;

            // Register events
            MinecraftForge.EVENT_BUS.register(this);
            Paragon.INSTANCE.getEventBus().register(this);
        }

        Arrays.stream(getClass().getDeclaredFields()).filter(field -> Setting.class.isAssignableFrom(field.getType())).forEach(field -> {
            field.setAccessible(true);

            try {
                Setting<?> setting = (Setting<?>) field.get(this);

                if (setting != null && setting.getParentSetting() == null) {
                    settings.add(setting);
                }
            } catch (IllegalAccessException | IllegalArgumentException e) {
                e.printStackTrace();
            }
        });

        this.settings.add(this.visible);
        this.settings.add(this.bind);
    }

    public Module(String name, Category category, String description, Bind bind) {
        this(name, category, description);

        this.bind.setValue(bind);
    }

    public void onEnable() {
    }

    public void onDisable() {
    }

    public void onTick() {
    }

    public void onRender2D() {
    }

    public void onRender3D() {
    }

    /**
     * Toggles the module
     */
    public void toggle() {
        // We don't want to toggle if the module is constant
        if (constant) {
            return;
        }

        this.enabled = !enabled;

        ModuleToggleEvent moduleToggleEvent = new ModuleToggleEvent(this);
        Paragon.INSTANCE.getEventBus().post(moduleToggleEvent);

        if (moduleToggleEvent.isCancelled()) {
            return;
        }

        if (enabled) {
            // Register events
            MinecraftForge.EVENT_BUS.register(this);
            Paragon.INSTANCE.getEventBus().register(this);

            animation.setState(true);

            // Call onEnable
            onEnable();
        } else {
            // Unregister events
            MinecraftForge.EVENT_BUS.unregister(this);
            Paragon.INSTANCE.getEventBus().unregister(this);

            animation.setState(false);

            // Call onDisable
            onDisable();
        }
    }

    /**
     * Gets the module info for the array list
     *
     * @return The module's info
     */
    public String getArrayListInfo() {
        return "";
    }

    /**
     * Gets the module's category
     *
     * @return The module's category
     */
    public Category getCategory() {
        return category;
    }

    /**
     * Gets whether the module is enabled
     *
     * @return Whether the module is enabled
     */
    public boolean isEnabled() {
        return enabled;
    }

    /**
     * Gets the module's visibility
     *
     * @return The module's visibility
     */
    public boolean isVisible() {
        return visible.getValue();
    }

    /**
     * Sets the module's visibility
     *
     * @param visible The module's new visibility
     */
    public void setVisible(boolean visible) {
        this.visible.setValue(visible);
    }

    /**
     * Checks if the module is constantly enabled
     *
     * @return Whether the module is constantly enabled
     */
    public boolean isConstant() {
        return constant;
    }

    /**
     * Gets whether the module is ignored by notifications or not
     *
     * @return Whether the module is ignored by notifications or not
     */
    public boolean isIgnored() {
        return ignoredByNotifications;
    }

    /**
     * Gets a list of the module's settings
     *
     * @return The module's settings
     */
    public List<Setting<?>> getSettings() {
        return settings;
    }

    /**
     * Gets the bind of the module
     *
     * @return The bind of the module
     */
    public Setting<Bind> getBind() {
        return bind;
    }
}

File: src/main/kotlin/com/paragon/api/module/Module.kt
package com.paragon.api.module

import com.paragon.Paragon
import com.paragon.api.event.client.ModuleToggleEvent
import com.paragon.api.feature.Feature
import com.paragon.api.setting.Bind
import com.paragon.api.setting.Setting
import com.paragon.api.util.Wrapper
import com.paragon.client.systems.module.hud.impl.ArrayListHUD
import me.surge.animation.Animation
import net.minecraftforge.common.MinecraftForge
import org.lwjgl.input.Keyboard
import java.util.*

open class Module(name: String, val category: Category, description: String) : Feature(name, description), Wrapper {

    // Whether the module is visible in the Array List or not
    private val visible = Setting("Visible", true).setDescription("Whether the module is visible in the array list or not")

    val bind = Setting("Bind", Bind(Keyboard.KEY_NONE, Bind.Device.KEYBOARD)).setDescription("The keybind of the module")

    // Whether the module is constantly enabled or not
    private val isConstant = javaClass.isAnnotationPresent(Constant::class.java)

    // Whether the module is ignored by notifications
    val isIgnored = javaClass.isAnnotationPresent(IgnoredByNotifications::class.java)

    // Module Settings
    val settings: MutableList<Setting<*>> = ArrayList()

    // Arraylist animation
    val animation = Animation({ ArrayListHUD.animationSpeed.value }, false) { ArrayListHUD.easing.value }

    // Whether the module is enabled
    var isEnabled = false
        private set

    init {
        if (isConstant) {
            isEnabled = true

            // Register events
            MinecraftForge.EVENT_BUS.register(this)
            Paragon.INSTANCE.eventBus.register(this)
        }
    }

    constructor(name: String, category: Category, description: String, bind: Bind) : this(name, category, description) {
        this.bind.setValue(bind)
    }

    // TEMPORARY
    fun reflectSettings() {
        Arrays.stream(javaClass.declaredFields)
            .filter { Setting::class.java.isAssignableFrom(it.type) }
            .forEach {
                it.isAccessible = true
                try {
                    val setting = it[this] as Setting<*>
                    if (setting.parentSetting == null) {
                        settings.add(setting)
                    }
                } catch (e: Throwable) {
                    e.printStackTrace()
                }
            }

        settings.add(visible)
        settings.add(bind)
    }

    open fun onEnable() {}
    open fun onDisable() {}
    open fun onTick() {}
    open fun onRender2D() {}
    open fun onRender3D() {}

    open fun getData() = ""

    open fun isActive() = isEnabled

    /**
     * Toggles the module
     */
    fun toggle() {
        // We don't want to toggle if the module is constant
        if (isConstant) {
            return
        }

        isEnabled = !isEnabled

        Paragon.INSTANCE.eventBus.post(ModuleToggleEvent(this))

        if (isEnabled) {
            // Register events
            MinecraftForge.EVENT_BUS.register(this)
            Paragon.INSTANCE.eventBus.register(this)
            animation.state = true

            // Call onEnable
            onEnable()
        } else {
            // Unregister events
            MinecraftForge.EVENT_BUS.unregister(this)
            Paragon.INSTANCE.eventBus.unregister(this)
            animation.state = false

            // Call onDisable
            onDisable()
        }
    }

    /**
     * Gets the module's visibility
     *
     * @return The module's visibility
     */
    fun isVisible() = visible.value

    /**
     * Sets the module's visibility
     *
     * @param visible The module's new visibility
     */
    fun setVisible(visible: Boolean) {
        this.visible.setValue(visible)
    }

}